src/library/rekordbox/rekordboxfeature.cpp

```diff
@@ -0,0 +1,977 @@
+    // is completed, this can be revisted.
+    // Attempt was made to also recover HISTORY
+    // playlists (which are found on removable Rekordbox devices), however
+    // they didn't appear to contain valid row_ref_t structures.
+    const int totalTables = 8;
+
+    rekordbox_pdb_t::page_type_t tableOrder[totalTables] = {
+            rekordbox_pdb_t::PAGE_TYPE_KEYS,
+            rekordbox_pdb_t::PAGE_TYPE_GENRES,
+            rekordbox_pdb_t::PAGE_TYPE_ARTISTS,
+            rekordbox_pdb_t::PAGE_TYPE_ALBUMS,
+            rekordbox_pdb_t::PAGE_TYPE_PLAYLIST_ENTRIES,
+            rekordbox_pdb_t::PAGE_TYPE_TRACKS,
+            rekordbox_pdb_t::PAGE_TYPE_PLAYLIST_TREE,
+            rekordbox_pdb_t::PAGE_TYPE_HISTORY};
+
+    QMap<uint32_t, QString> keysMap;
+    QMap<uint32_t, QString> genresMap;
+    QMap<uint32_t, QString> artistsMap;
+    QMap<uint32_t, QString> albumsMap;
+    QMap<uint32_t, QString> playlistNameMap;
+    QMap<uint32_t, bool> playlistIsFolderMap;
+    QMap<uint32_t, QMap<uint32_t, uint32_t>> playlistTreeMap;
+    QMap<uint32_t, QMap<uint32_t, uint32_t>> playlistTrackMap;
+
+    bool folderOrPlaylistFound = false;
+
+    for (int tableOrderIndex = 0; tableOrderIndex < totalTables; tableOrderIndex++) {
+        bool done = false;
+
+        for (
+                std::vector<rekordbox_pdb_t::table_t *>::iterator table = reckordboxDB.tables()->begin();
+                table != reckordboxDB.tables()->end();
+                ++table) {
+            if ((*table)->type() == tableOrder[tableOrderIndex]) {
+                if (done)
+                    break;
+
+                uint16_t lastIndex = (*table)->last_page()->index();
+                rekordbox_pdb_t::page_ref_t *currentRef = (*table)->first_page();
+                bool moreLeft = true;
+
+                do {
+                    rekordbox_pdb_t::page_t *page = currentRef->body();
+
+                    if (page->is_data_page()) {
+                        for (
+                                std::vector<rekordbox_pdb_t::row_group_t *>::iterator rowGroup = page->row_groups()->begin();
+                                rowGroup != page->row_groups()->end();
+                                ++rowGroup) {
+                            for (
+                                    std::vector<rekordbox_pdb_t::row_ref_t *>::iterator rowRef = (*rowGroup)->rows()->begin();
+                                    rowRef != (*rowGroup)->rows()->end();
+                                    ++rowRef) {
+                                if ((*rowRef)->present()) {
+                                    switch (tableOrder[tableOrderIndex]) {
+                                    case rekordbox_pdb_t::PAGE_TYPE_KEYS: {
+                                        // Key found, update map
+                                        rekordbox_pdb_t::key_row_t *key =
+                                                (rekordbox_pdb_t::key_row_t *)(*rowRef)->body();
```

Comment:
Please replace all those remaining C-style casts with the appropriate C++-cast, i.e. `static_cast<>`